Utilities/MCStepLogger/src/AnalysisBasic.cxx

```diff
@@ -0,0 +1,281 @@
+      // check in how many events a certain volume was present
+      if (pdgPresent.find(pdgString) == pdgPresent.end()) {
+        pdgPresent.insert(std::pair<std::string, float>(pdgString, 1.));
+      } else {
+        pdgPresent[pdgString]++;
+      }
+    }
+  }
+
+  void finalize() override
+  {
+    // fill and update (e.g. scaling) some histograms
+    histNVols->Fill(0.5, volIds.size());
+
+    // just normalise over all events
+    histNStepsPerEvent->Scale(1. / histNEvents->GetEntries());
+    histNTracksPerEvent->Scale(1. / histNEvents->GetEntries());
+    histNSecondariesPerEvent->Scale(1. / histNEvents->GetEntries());
+    histStepsXPerEvent->Scale(1. / histNEvents->GetEntries());
+    histStepsYPerEvent->Scale(1. / histNEvents->GetEntries());
+    histStepsZPerEvent->Scale(1. / histNEvents->GetEntries());
+    histStepSizesPerEvent->Scale(1. / histNEvents->GetEntries());
+    histMeanStepSizePerEvent->Scale(1. / histNEvents->GetEntries());
+    // normalize per volume to number of events where a certain volume was present
+    for (auto& vp : volPresent) {
+      vp.second = 1. / vp.second;
+    }
+    for (auto& pp : pdgPresent) {
+      pp.second = 1. / pp.second;
+    }
+    // scale histograms bin-wise
+    AnalysisHelper::scalePerBin(histNStepsPerVolPerEvent, volPresent);
+    AnalysisHelper::scalePerBin(histMagFieldCallsPerVolPerEvent, volPresent);
+    AnalysisHelper::scalePerBin(histSmallMagFieldCallsPerVolPerEvent, volPresent);
+    AnalysisHelper::scalePerBin(histNSecondariesPerVolPerEvent, volPresent);
+    AnalysisHelper::scalePerBin(histMeanStepSizePerVolPerEvent, volPresent);
+    AnalysisHelper::scalePerBin(histMeanStepSizePerPDGPerEvent, pdgPresent);
+    AnalysisHelper::scalePerBin(histNStepsPerPDGPerEvent, pdgPresent);
+    AnalysisHelper::scalePerBin(histNTracksPerPDGPerEvent, pdgPresent);
+
+    int nVolBins = histNStepsPerVolPerEvent->GetNbinsX();
+    for (int i = 1; i < nVolBins + 1; i++) {
+      histRelNStepsPerVolPerEvent->Fill(histNStepsPerVolPerEvent->GetXaxis()->GetBinLabel(i), histNStepsPerVolPerEvent->GetBinContent(i));
+    }
+    histRelNStepsPerVolPerEvent->Scale(1. / double(histNStepsPerEvent->GetMaximum()));
+
+    int nPDGBins = histNTracksPerPDGPerEvent->GetNbinsX();
+    for (int i = 1; i < nPDGBins + 1; i++) {
+      histRelNTracksPerPDGPerEvent->Fill(histNTracksPerPDGPerEvent->GetXaxis()->GetBinLabel(i), histNTracksPerPDGPerEvent->GetBinContent(i));
+    }
+    histRelNTracksPerPDGPerEvent->Scale(1. / double(histNTracksPerEvent->GetMaximum()));
+
+    nPDGBins = histNStepsPerPDGPerEvent->GetNbinsX();
+    for (int i = 1; i < nPDGBins + 1; i++) {
+      histRelNStepsPerPDGPerEvent->Fill(histNStepsPerPDGPerEvent->GetXaxis()->GetBinLabel(i), histNStepsPerPDGPerEvent->GetBinContent(i));
+    }
+    histRelNStepsPerPDGPerEvent->Scale(1. / double(histNStepsPerEvent->GetMaximum()));
+  }
+
+ private:
+  TH1I* histNEvents;
```

Comment:
pleas add docu for this data! Do you need pointers or can objects work as well?

Reply:
These are pointers since they are actually owned by an AnalysisFile object handling the lifetime of these objects